### PR TITLE
Enforce Dataframe Backend Checks

### PR DESCRIPTION
**Pull request by ryantwolf**
## Description

Refactors all modules to inherit from `nemo_curator.modules.base.Module`. With this change, the base class now checks to see if the dataframe backend for the `Module` matches what is passed in, and will throw an error prior to computation if there is a mismatch. We also prove a new module `ToBackend()` that allows easy chaining of CPU and GPU modules.

## Usage

```python
from nemo_curator import Module, Sequential, ToBackend
from nemo_curator.datasets import DocumentDataset

class CPUModule(Module):
    def __init__(self):
        super().__init__(input_backend="pandas")

    def call(self, dataset: DocumentDataset):
        dataset.df["cpu_lengths"] = dataset.df["text"].str.len()
        return dataset


class GPUModule(Module):
    def __init__(self):
        super().__init__(input_backend="cudf")

    def call(self, dataset: DocumentDataset):
        dataset.df["gpu_lengths"] = dataset.df["text"].str.len()
        return dataset

def main():
  client = get_client(cluster_type="gpu")
  # Load a dataset with a pandas CPU backend
  dataset = DocumentDataset.from_pandas(...)
  pipeline = Sequential([
    CPUModule(),
    ToBackend("cudf"),
    GPUModule(),
  ])

  curated_dataset = pipeline(dataset)
  curated_dataset.to_json(...)

if __name__ == "__main__":
  main()
```
## Checklist

- [x] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [x] New or Existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: tests/test_backends.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+from nemo_curator import Module, Sequential, ToBackend
+from nemo_curator.datasets import DocumentDataset
+from nemo_curator.utils.import_utils import gpu_only_import, gpu_only_import_from
+
+cudf = gpu_only_import("cudf")
+dask_cudf = gpu_only_import("dask_cudf")
+LocalCUDACluster = gpu_only_import_from("dask_cuda", "LocalCUDACluster")
+
+
+class CPUModule(Module):
```
Comment (sarahyurick) [inline tests/test_backends.py:28]: Was wondering if we can also have some type of test with at least 1 CPU and GPU module? Such as fuzzy duplicates?

File: docs/user-guide/cpuvsgpu.rst [hunk 2]
```diff
@@ -85,6 +85,44 @@ To read a dataset into GPU memory, one could use the following function call.
...
+
+  def main():
+      client = get_client(cluster_type="gpu")
+
+      dataset = DocumentDataset.read_json("books.jsonl")
+      curation_pipeline = Sequential([
+          ScoreFilter(RepeatingTopNGramsFilter(n=5)),
+          ToBackend("cudf"),
+          DomainClassifier(),
+      ])
```
Comment (sarahyurick) [inline docs/user-guide/cpuvsgpu.rst:110]: This is very nice!

File: nemo_curator/modules/__init__.py [hunk 3]
```diff
@@ -22,12 +22,14 @@
...
 from .config import FuzzyDuplicatesConfig, SemDedupConfig
 from .dataset_ops import blend_datasets, Shuffle
 from .exact_dedup import ExactDuplicates
 from .meta import Sequential
 from .modify import Modify
 from .task import TaskDecontamination
+from .to_backend import ToBackend
 
 # GPU packages
 MinHash = gpu_only_import_from("nemo_curator.modules.fuzzy_dedup.minhash", "MinHash")
```
Comment (sarahyurick) [inline nemo_curator/modules/__init__.py:35]: Do you think there is any way to take advantage of the module backends so that we don't need to use `gpu_only_import_from`?

File: tests/test_backends.py (hunk 1 above)
Comment (ryantwolf) [reply tests/test_backends.py:28]: you mean with a real module, not just these dummy modules?

File: nemo_curator/modules/__init__.py (hunk 3 above)
Comment (ryantwolf) [reply nemo_curator/modules/__init__.py:35]: Not quite sure what you mean. Are you asking if we could somehow determine that `MinHash` is a GPU module from its backend and then refactor this to do something different than `gpu_only_import_from`? I'm wondering what you think should be improved with these imports.

File: tests/test_backends.py (hunk 1 above)
Comment (sarahyurick) [reply tests/test_backends.py:28]: Yes.

File: docs/user-guide/cpuvsgpu.rst [hunk 4]
```diff
@@ -85,6 +85,44 @@ To read a dataset into GPU memory, one could use the following function call.
...
+
+      curated_dataset.to_json("curated_books.jsonl")
+
+  if __name__ == "__main__":
+      main()
+
+Let's highlight some of the important parts of this example.
+
+* ``client = get_client(cluster_type="gpu")``: Creates a local Dask cluster with access to the GPUs. In order to use/swap to a cuDF dataframe backend, you need to make sure you are running on a GPU Dask cluster.
+* ``dataset = DocumentDataset.read_json("books.jsonl")``: Reads in the dataset to a pandas (CPU) backend.
```
Comment (praateekmahajan) [inline docs/user-guide/cpuvsgpu.rst:122]: I think we should I explicitly mention the implicit default argument of `backend="pandas"`

File: nemo_curator/filters/doc_filter.py [hunk 5]
```diff
@@ -81,6 +81,15 @@ def keep_document(self, scores: Any) -> bool:
             "keep_document method must be implemented by subclasses"
         )
 
+    def get_backend(self) -> str:
```
Comment (praateekmahajan) [inline nemo_curator/filters/doc_filter.py:84]:
```suggestion
    def get_backend(self) -> Literal["pandas", "cudf", "any"]:
```

(what's any though?)

File: nemo_curator/modules/base.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    @abstractmethod
+    def call(self, dataset: DocumentDataset) -> DocumentDataset:
+        raise NotImplementedError("call method must be implemented by subclasses")
+
+    def _check_backend(self, ddf):
+        if self.input_backend == "any":
+            return
+
+        backend = type(ddf._meta).__module__.split(".")[0]
```
Comment (praateekmahajan) [inline nemo_curator/modules/base.py:41]: Can we add a test case for this? And also type hints for ddf which would be `dd.DataFrame`

File: nemo_curator/modules/__init__.py (hunk 3 above)
Comment (sarahyurick) [reply nemo_curator/modules/__init__.py:35]: I guess I was wondering if it would be possible to use the backend checks in this PR to help all modules be importable by users. For example if a user with CPU only tries to do an `from nemo_curator import FuzzyDuplicates` then maybe this file should raise a more readable message which specifies that the module can only be run with the GPU?

File: nemo_curator/modules/base.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            raise ValueError(
+                f"{input_backend} not one of the supported backends {self.SUPPORTED_BACKENDS}"
+            )
+        self.input_backend = input_backend
+
+    @abstractmethod
+    def call(self, dataset: DocumentDataset) -> DocumentDataset:
+        raise NotImplementedError("call method must be implemented by subclasses")
+
+    def _check_backend(self, ddf):
```
Comment (praateekmahajan) [inline nemo_curator/modules/base.py:37]: nit : name could be `_assert_correct_backend`

File: nemo_curator/filters/doc_filter.py (hunk 5 above)
Comment (praateekmahajan) [reply nemo_curator/filters/doc_filter.py:84]: Seems like this could be an [abstractproperty](https://docs.python.org/3/library/abc.html#abc.abstractproperty) (now abstractmethod + property). 

And instead of the abstract class defining it, IMO the inherited class should explicitly define that property to be `pandas` or whatever

File: nemo_curator/modifiers/doc_modifier.py [hunk 8]
```diff
@@ -26,3 +26,12 @@ def __init__(self):
...
         pass
+
+    def get_backend(self) -> str:
+        """
+        The dataframe backend the filter operates on.
+        Can be 'pandas', 'cudf', or 'any'. Defaults to 'pandas'.
+        Returns:
+            str: A string representing the dataframe backend the filter needs as input
+        """
+        return "pandas"
```
Comment (praateekmahajan) [inline nemo_curator/modifiers/doc_modifier.py:37]: Same as above on type hint + abstractproperty + require explicit definition in child class rather than abstract class

File: nemo_curator/modules/to_backend.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+from nemo_curator.modules.base import Module
+
+
+class ToBackend(Module):
+    def __init__(self, backend: str) -> None:
+        super().__init__(input_backend="any")
+        self.backend = backend
+
+    def call(self, dataset: DocumentDataset) -> DocumentDataset:
+        return DocumentDataset(dataset.df.to_backend(self.backend))
```
Comment (praateekmahajan) [inline nemo_curator/modules/to_backend.py:25]: can add an explicit no-op if backend is the same if it's easy to check

File: nemo_curator/modules/to_backend.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+# limitations under the License.
+
+from nemo_curator.datasets.doc_dataset import DocumentDataset
+from nemo_curator.modules.base import Module
+
+
+class ToBackend(Module):
+    def __init__(self, backend: str) -> None:
+        super().__init__(input_backend="any")
+        self.backend = backend
```
Comment (praateekmahajan) [inline nemo_curator/modules/to_backend.py:22]: nit : type hint

File: nemo_curator/modules/base.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        self.name = name or self.__class__.__name__
+
+        if input_backend not in self.SUPPORTED_BACKENDS:
+            raise ValueError(
+                f"{input_backend} not one of the supported backends {self.SUPPORTED_BACKENDS}"
+            )
+        self.input_backend = input_backend
+
+    @abstractmethod
+    def call(self, dataset: DocumentDataset) -> DocumentDataset:
```
Comment (praateekmahajan) [inline nemo_curator/modules/base.py:34]: super nit : `call` could be slightly confusing for python readers, as it does overlap with the `__call__` dunder. What about we call it `fn` / `func` / `impl` / `implementation`?

File: nemo_curator/modules/base.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+from abc import ABC, abstractmethod
+
+from nemo_curator.datasets import DocumentDataset
+
+
+class Module(ABC):
+    SUPPORTED_BACKENDS = ["pandas", "cudf", "any"]
+
+    def __init__(self, input_backend: str, name=None) -> None:
```
Comment (praateekmahajan) [inline nemo_curator/modules/base.py:23]: nit : type hints

File: tests/test_backends.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+                ToBackend("cudf"),
+                GPUModule(),
+                ToBackend("pandas"),
+                CPUModule(),
+                ToBackend("cudf"),
+                GPUModule(),
+            ]
+        )
+        result = pipeline(dataset)
+        result_df = result.df.compute()
```
Comment (praateekmahajan) [inline tests/test_backends.py:212]: add a small assert here for list(df.columns) which should be `id, text, gpu_lengths, cpu_lengths`

**Review by praateekmahajan — CHANGES_REQUESTED** [review]
This looks great. Left a review, with mostly nits around type hints and function renaming  and some class abstraction opinions.

File: nemo_curator/modules/__init__.py (hunk 3 above)
Comment (ryantwolf) [reply nemo_curator/modules/__init__.py:35]: I don't think it can. In `FuzzyDuplicates` for example, the import would fail before the `FuzzyDuplicates` class would be called due to the missing `dask_cudf` import. And we'd need the `FuzzyDuplicates` class to be there in order to check its backend.

File: nemo_curator/modules/base.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+
+from nemo_curator.datasets import DocumentDataset
+
+
+class Module(ABC):
```
Comment (praateekmahajan) [inline nemo_curator/modules/base.py:20]: nit to rename this to `BaseModule`

File: nemo_curator/filters/doc_filter.py (hunk 5 above)
Comment (ryantwolf) [reply nemo_curator/filters/doc_filter.py:84]: "any" means that the module can operate on either "pands" or "cudf" as a backend. Will update to be a property instead of a method.

> And instead of the abstract class defining it, IMO the inherited class should explicitly define that property to be pandas or whatever

The reason that I have the base class start off with a default of `pandas` is that right now I believe every DocumentFilter has a backend of pandas. In my mind, it's more convenient to set the default and have a few classes override it in the future if they want. What do you think?

File: nemo_curator/modules/base.py (hunk 6 above)
Comment (ryantwolf) [reply nemo_curator/modules/base.py:41]: What part of this do you want to test? I'm not quite seeing which part of this you want me to isolate.

File: nemo_curator/modules/base.py (hunk 7 above)
Comment (ryantwolf) [reply nemo_curator/modules/base.py:37]: what do you think about `_validate_correct_backend` instead? Since there's no direct `assert` statement or error thrown I think something slightly more generic is better.

File: nemo_curator/modules/base.py [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+        if self.input_backend == "any":
+            return
+
+        backend = type(ddf._meta).__module__.split(".")[0]
+        if backend != self.input_backend:
+            raise ValueError(
+                f"Module {self.name} requires dataset to have backend {self.input_backend} but got backend {backend}"
+            )
+
+    def __call__(self, dataset: DocumentDataset) -> DocumentDataset:
```
[the remark this answers is not in the document]
Comment (ryantwolf) [reply nemo_curator/modules/base.py:47]: I'll leave the return type unannotated for now.

File: nemo_curator/modules/to_backend.py (hunk 9 above)
Comment (ryantwolf) [reply nemo_curator/modules/to_backend.py:25]: From what I can see dask already has a no-op in its implementation for pandas, I assume the cudf one is similar. I'll keep it like this for now.

File: nemo_curator/modules/base.py (hunk 11 above)
Comment (sarahyurick) [reply nemo_curator/modules/base.py:34]: Hmm maybe something like `run`, `execute`, `dispatch`? Something that still makes it clear that we are calling the core functionality of that class.

**Comment by ryantwolf** [comment]
@sarahyurick @praateekmahajan thanks for your reviews. I addressed the feedback that I immediately agreed with/understood. I left some follow up comments and questions on the other points.

File: docs/user-guide/cpuvsgpu.rst [hunk 16]
```diff
@@ -85,6 +85,44 @@ To read a dataset into GPU memory, one could use the following function call.
...
+  from nemo_curator.filters import RepeatingTopNGramsFilter
+
+  def main():
+      client = get_client(cluster_type="gpu")
+
+      dataset = DocumentDataset.read_json("books.jsonl")
+      curation_pipeline = Sequential([
+          ScoreFilter(RepeatingTopNGramsFilter(n=5)),
+          ToBackend("cudf"),
+          DomainClassifier(),
```
Comment (ayushdg) [inline docs/user-guide/cpuvsgpu.rst:109]: It might be worth showing an example going the other way from cuDF to pandas via ToBackend in this short example as well.

File: nemo_curator/modules/base.py [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+        Args:
+            dataset (DocumentDataset): The dataset to operate on
+        """
+        raise NotImplementedError("call method must be implemented by subclasses")
+
+    def _validate_correct_backend(self, ddf: dd.DataFrame):
+        if self.input_backend == "any":
+            return
+
+        backend = type(ddf._meta).__module__.split(".")[0]
```
Comment (ayushdg) [inline nemo_curator/modules/base.py:67]: Would the `is_cudf_type` utility help in this case? https://github.com/NVIDIA/NeMo-Curator/blob/c3fb61dfc37d25a916384c8484a8da2252e9e105/nemo_curator/utils/gpu_utils.py#L19-L28

File: nemo_curator/modules/base.py [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+        if backend != self.input_backend:
+            raise ValueError(
+                f"Module {self.name} requires dataset to have backend {self.input_backend} but got backend {backend}"
+            )
+
```
Comment (ayushdg) [inline nemo_curator/modules/base.py:72]: In addition to `_validate_correct_backend`, it also makes sense to have something like `_validate_correct_cluster_type` that can also verify that for cuDF backends folks are using a GPU cluster. Perhaps the `getClient` api can set the cluster type in some variable? 

I realize that there is no clean way to do this today.

File: nemo_curator/modules/base.py [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+        raise NotImplementedError("call method must be implemented by subclasses")
+
+    def _validate_correct_backend(self, ddf: dd.DataFrame):
+        if self.input_backend == "any":
+            return
+
+        backend = type(ddf._meta).__module__.split(".")[0]
+        if backend != self.input_backend:
+            raise ValueError(
+                f"Module {self.name} requires dataset to have backend {self.input_backend} but got backend {backend}"
```
Comment (ayushdg) [inline nemo_curator/modules/base.py:70]: Might be worth giving a hint here by mentioning that Users can use the `ToBackend` module to switch backends.

File: nemo_curator/modules/to_backend.py [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+
+    def __init__(self, backend: Literal["pandas", "cudf"]) -> None:
+        """
+        Constructs a ToBackend module
+
+        Args:
+            backend (str): The backend to transfer the dataset to. Can be "pandas" or "cudf"
+        """
+        super().__init__(input_backend="any")
+        self.backend = backend
```
Comment (ayushdg) [inline nemo_curator/modules/to_backend.py:33]: Might be worth adding the `supported_backends` list and check here as well to ensure users don't try to convert to an unsupported backend.

File: tests/test_backends.py [hunk 21]
```diff
@@ -0,0 +1,437 @@
...
+        return dataset
+
+
+class AnyModule(BaseModule):
+    def __init__(self):
+        super().__init__(input_backend="any")
+
+    def call(self, dataset: DocumentDataset):
+        dataset.df["any_lengths"] = dataset.df["text"].str.len()
+        return dataset
```
Comment (ayushdg) [inline tests/test_backends.py:69]: repeated class

File: tests/test_backends.py [hunk 22]
```diff
@@ -0,0 +1,437 @@
...
+
+
+@pytest.fixture
+def cpu_data():
+    df = pd.DataFrame(
```
Comment (ayushdg) [inline tests/test_backends.py:74]: nit: You could create a common fixture that returns the original data and have two fixtures that use the common one as input for generating cpu and gpu data instead of repeating the data twice.

eg: 
```python
@pytest.fixture
def raw_data():
...

@pytest.fixture
def cpu_data(raw_data):
    return ...

@pytest.fixture
def gpu_data(raw_data):
 ...
```

File: tests/test_backends.py [hunk 23]
```diff
@@ -0,0 +1,437 @@
...
+        assert sorted(list(result_df.columns)) == [
+            "cpu_lengths",
+            "gpu_lengths",
+            "id",
+            "text",
+        ]
+        assert_eq(result_df["cpu_lengths"], gt_cpu_lengths)
+        assert_eq(result_df["gpu_lengths"], gt_gpu_lengths)
+
+    def test_wrong_backend(self, cpu_data):
```
Comment (ayushdg) [inline tests/test_backends.py:239]: nit: could you add the other way around as well?

File: tests/test_backends.py [hunk 24]
```diff
@@ -0,0 +1,437 @@
...
+        with pytest.raises(ValueError):
+            print("client", self.client)
+            dataset, _ = cpu_data
+            pipeline = GPUModule()
+            result = pipeline(dataset)
+            _ = result.df.compute()
+
+
+@pytest.fixture
```
Comment (ayushdg) [inline tests/test_backends.py:248]: same as above

**Review by ayushdg — CHANGES_REQUESTED** [review]

File: nemo_curator/modules/base.py (hunk 11 above)
Comment (ayushdg) [reply nemo_curator/modules/base.py]: No strong opinions: open to using `run/execute/call`.

Comment (VibhuJawa) [reply nemo_curator/modules/base.py]: My opinion is actually to stick with `call`.  I think CrossFit call convention comes from Keras/TensorFlow (lol) Anyways, my mental model is call actually ends up calling `__call__ ` function so it lines up in my head. 

Other options can be `transform`, `process`  too.  

I slightly dont like `run`, `execute` because it has annotations of synchronous behavior in my head (but can be convinced of it too).

Comment (ryantwolf) [reply nemo_curator/modules/base.py]: Ok given that it's from keras/tensorflow and crossfit I'm going to keep it that way for now.

File: nemo_curator/modules/base.py (hunk 18 above)
Comment (ryantwolf) [reply nemo_curator/modules/base.py:72]: I think it does make sense to do this. We could have a global variable `CLUSTER_TYPE` that gets set during the `get_client` call. Then, in the `BaseModule` we can just check this variable. The only problem that I see with this right now though is how to handle the case where we connect to an existing Dask cluster. We'd need to make sure that it also set some environment variable or something designating the cluster type. I think we can enforce this in our example slurm scripts, but do you mind if I do this in another PR?

Comment (ayushdg) [reply nemo_curator/modules/base.py:72]: Makes sense. Let's handle this in a followup. For cases where this variable is not set like when users are setting things up themselves without using our recommended scripts, we can keep the check lenient enough to still allow processing (maybe with a warning if preferred).

File: tests/test_backends.py (hunk 22 above)
Comment (ryantwolf) [reply tests/test_backends.py:74]: sweet, thanks for letting me know!

**Review by sarahyurick — APPROVED** [review]
LGTM!

File: nemo_curator/modules/base.py (hunk 6 above)
Comment (praateekmahajan) [reply nemo_curator/modules/base.py]: As someone reading this, it's not obviously clear what will be the output of `.__module__.split(".")[0]` of the type of  object's `_meta` attribute. So I would've proposed that you move this to a function and have tests for that and use that function here. 
However I see you've changed this to is_cudf_type(..) which is much more readable, and now we can unittest that directly that.

**Review by praateekmahajan — APPROVED** [review]
LGTM thanks for adding this!

**Review by ayushdg — APPROVED** [review]
lgtm! Thanks
